### ISSUE TITLE
Issue when reproducing iSpikformer results on Electricity dataset

### ISSUE THREAD

**Issue by huangrenzhuo** [microsoft/SeqSNN#24]
Hello, I'm trying to reproduce the results of iSpikformer as described in the paper "Efficient and Effective Time-Series Forecasting with Spiking Neural Networks", but my experimental results on the Electricity dataset differ significantly from those reported. I noticed the paper mentions averaging results across 3 random seeds — could you clarify which specific seeds were used in the original experiments?

Below are my configuration details and outcomes
```
_base_:
- ../dataset/electricity.yml
_custom_imports_:
- SeqSNN.network
- SeqSNN.dataset
- SeqSNN.runner

data:
  raw_label: False
  window: 168
  horizon: 24
  normalize: 3

runner:
  type: vsts
  task: regression
  optimizer: Adam
  lr: 0.0001
  weight_decay: 0.0
  loss_fn: mse
  metrics: [r2, rrse]
  observe: loss
  batch_size: 32
  out_size: 24
  aggregate: False
  early_stop: 30
  denormalize: True
  lower_is_better: True
  max_epoches: 1000
  valid_variates: 321

network:
  type: iSpikformer
  dim: 256
  d_ff: 1024
  depths: 2
  num_steps: 4
  heads: 8
  encoder_type: conv
```

![Image](https://github.com/user-attachments/assets/88d47834-0fb4-40f6-8c06-0ade0bdcdacc)

**Comment by Lvchangze** [comment]
It seems that **your results significantly outperform ours**.
I am not sure which Linux version and seed you used. 
We have obtained all the model results in a fair way and with reasonable limits. There may be some parameter combinations that can achieve better results on some model and dataset combinations.
We are also willing to further discuss these results and hope to provide your experimental parameters, especially the computer system version and seed value.

You can set your seed configuration as follows:
runtime:
  seed: 42

**Comment by huangrenzhuo** [comment]
Thank you very much for your reply.

I'd also like to provide my environment details to help troubleshoot the issue:
* **OS**: Ubuntu 22.04.5 LTS
* **Random Seed**: 42

After reviewing your response, I went back through the paper, the supplementary material, and the code repository. I found some apparent inconsistencies regarding the model's parameter settings and was hoping you could clarify them.

1.  **From the Supplementary Material**: The supplement states that for all iSpikeformer models, the `dim` is set to 512 and `d_ff` is set to 1024.

2.  **Contradiction with the Paper's Parameter Count**: The main paper mentions that the iSpikeformer model for the Electricity dataset has 1.63M parameters. However, I calculated that the parameter count is only around 1.63M when `dim` is set to 256 and `d_ff` is set to 1024. If `dim` were 512 as suggested in the supplement, the parameter count would be much higher.

3.  **Inconsistencies in the GitHub Repository**: I also noticed that the default scripts in the GitHub repository differ from the supplementary material's description in a couple of ways.
    * The repository specifies different `dim` and `d_ff` settings for different datasets (`dim`=256 and `d_ff`=1024 for electricity & metr-la, but `dim`=512 and `d_ff`=2048 for solar & pems-bay).
    * Additionally, the supplementary material indicates that the SSA `threshold` is 0.25, but I could not find where this is set in the provided code.

Based on these points, I am a bit confused. To ensure I can correctly reproduce your experimental results, could you please clarify the following?

* What were the actual `dim`, `d_ff`, and SSA `threshold` configurations for the iSpikeformer on the four datasets (electricity, metr-la, solar, and pems-bay) in the paper's experiments?
* Were the model configurations intended to be the same across all four datasets, or do they vary as suggested by the code repository?

Thank you again for your help!

**Comment by Lvchangze** [comment]
Yes, the parameters in the configuration file in the GitHub repository are not the configuration files reported in our appendix.
In fact, we followed the hyperparameter search settings of the previous article [1] and performed a hyperparameter search.
For example, we set the range of each parameter, such as dim=[128, 256, 512], layer=[1, 2, 4], batch_size=[96, 128], seed=[41, 42, 43] and other hyperparameter combinations, and we selected the several parameter combinations that achieved the satisfying results.
Hyperparameter search is a very common method to improve model performance.
It seems that your setting of d_model=256, d_ff=1024, layer=2 can achieve better performance than our settings.
Maybe you should conduct a hyperparameter search under your computer configuration.

[1] Learning Decomposed Spatial Relations for Multi-Variate Time-Series Modeling. AAAI 2023

**Comment by huangrenzhuo** [comment]
Thank you very much for your patient responses. During the reproduction process, I encountered several other issues and hope to get your patient clarification.  


1. Using the default configuration in `ispikformer_metr-la.yml`, I initially encountered the same problem as described in [issue #10](https://github.com/microsoft/SeqSNN/issues/10). With the default settings provided in the GitHub repository, I obtained identical results to that issue. After careful inspection, I noticed that `encoder_type` in `ispikformer_metr-la.yml` was set to "delta". When I changed it to "conv", I successfully reproduced the results for horizons 6 and 24. However, I still cannot reproduce the results for horizons 48 and 96—the reproduced results are significantly different from those reported in your paper. The same issue occurs with the PEMS-BAY dataset.  

I also noticed that both datasets use a window size of 12 in the experiments. However, when reviewing the paper *TS-LIF: A TEMPORAL SEGMENT SPIKING NEURON NETWORK FOR TIME SERIES FORECASTING* (which takes your work as a baseline), they used a window size of 168 for their predictions. I wonder if the default experimental settings for these two datasets might have partial issues, similar to [issue #10](https://github.com/microsoft/SeqSNN/issues/10) in the GitHub repository?  

<img width="2133" height="864" alt="Image" src="https://github.com/user-attachments/assets/24d16c40-bbde-467f-b0a6-890aad33d398" />


2. I encountered problems while reproducing Spike-RNN and Spike-GRU as well:  

- For Spike-RNN, the default configuration uses `network: SpikeRNN` and `runner: ts`. Despite following the default settings, I failed to reproduce the results reported in the paper. Additionally, I noticed another network `SpikeRNN2D`, which mentions "time step" in its comments, but the `time step` variable is not actually used in the code implementation.  

- For Spike-GRU, the default configuration specifies `network: ISNNGRU2d` and `runner: snnvsts`. However, `snnvsts` is not implemented in the code, and `ISNNGRU2d` is not imported in `SeqSNN.SeqSNN.runner.__init__.py`.  

My understanding is that the `vsts` runner is only required for models using inverse embedding. Does the "I" in `ISNNGRU2d` stand for "inverse"? Also, since GRU is a type of RNN, why do Spike-GRU and Spike-RNN use different runners? Their networks also differ (one is 2D, the other is not)—could you clarify this?  


If possible, could you share the correct experimental settings for these models? I apologize for disturbing you and greatly appreciate your help!
